### PR TITLE
[ISSUE-161] base flow of drive replacement process

### PR DESCRIPTION
**Pull request by mishoyama**
## Purpose
[ISSUE-161](https://github.com/dell/csi-baremetal/issues/161)

## Steps to test

-  Deploy test app and check `drives` and `volumes`
```
$ drives
Id                                     Size        Type   SN                   Health   Usage    Status   Node
433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   104857600   HDD    LOOPBACK3898502982   GOOD     IN_USE   ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dc5e6b7-3f38-4515-b420-5d003d614e34   104857600   HDD    LOOPBACK2176849526   GOOD     IN_USE   ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dcbd70a-02d6-4f25-96cc-1c4f66932972   104857600   HDD    LOOPBACK4005496940   GOOD     IN_USE   ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
$ volumes
Id                                         Size        Type   Health   OpStatus    CSIStatus   Usage    Location                               Node
pvc-3d5078a7-e362-4de0-872c-274d6d75c0e3   104857600   xfs    GOOD     OPERATIVE   PUBLISHED   IN_USE   6dc5e6b7-3f38-4515-b420-5d003d614e34   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-48a07884-b021-4b0e-a270-b714c30f9a03   104857600   xfs    GOOD     OPERATIVE   PUBLISHED   IN_USE   6dcbd70a-02d6-4f25-96cc-1c4f66932972   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-84dc45d0-8a29-40ba-9fe8-0fee72dd0768   104857600   xfs    GOOD     OPERATIVE   PUBLISHED   IN_USE   433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   6ada4e6d-861a-4454-8c86-71d24be4f992
```
- Fail one drive
```
$ cat loopback.yaml 
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: default
  name: loopback-config
  labels:
    app: baremetal-csi-node
data:
  config.yaml: |-
    defaultDrivePerNodeCount: 3
    defaultDriveSize: 100Mi
    nodes:
    - nodeID: kind-worker 
      drives:
        - serialNumber: LOOPBACK3898502982 
          health: BAD
$ kubectl apply -f loopback.yaml 
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
configmap/loopback-config configured
$ drives
Id                                     Size        Type   SN                   Health   Usage       Status   Node
433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   105906176   HDD    LOOPBACK3898502982   BAD      RELEASING   ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dc5e6b7-3f38-4515-b420-5d003d614e34   104857600   HDD    LOOPBACK2176849526   GOOD     IN_USE      ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dcbd70a-02d6-4f25-96cc-1c4f66932972   104857600   HDD    LOOPBACK4005496940   GOOD     IN_USE      ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
$ volumes
Id                                         Size        Type   Health   OpStatus    CSIStatus   Usage       Location                               Node
pvc-3d5078a7-e362-4de0-872c-274d6d75c0e3   104857600   xfs    GOOD     OPERATIVE   PUBLISHED   IN_USE      6dc5e6b7-3f38-4515-b420-5d003d614e34   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-48a07884-b021-4b0e-a270-b714c30f9a03   104857600   xfs    GOOD     OPERATIVE   PUBLISHED   IN_USE      6dcbd70a-02d6-4f25-96cc-1c4f66932972   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-84dc45d0-8a29-40ba-9fe8-0fee72dd0768   104857600   xfs    BAD      OPERATIVE   PUBLISHED   RELEASING   433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   6ada4e6d-861a-4454-8c86-71d24be4f992
```
- Mark `volume `as released
```
$ kubectl annotate volume pvc-84dc45d0-8a29-40ba-9fe8-0fee72dd0768 release=done
volume.baremetal-csi.dellemc.com/pvc-84dc45d0-8a29-40ba-9fe8-0fee72dd0768 annotated
$ volumes
Id                                         Size        Type   Health   OpStatus    CSIStatus   Usage      Location                               Node
pvc-3d5078a7-e362-4de0-872c-274d6d75c0e3   104857600   xfs    GOOD     OPERATIVE   PUBLISHED   IN_USE     6dc5e6b7-3f38-4515-b420-5d003d614e34   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-48a07884-b021-4b0e-a270-b714c30f9a03   104857600   xfs    GOOD     OPERATIVE   PUBLISHED   IN_USE     6dcbd70a-02d6-4f25-96cc-1c4f66932972   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-84dc45d0-8a29-40ba-9fe8-0fee72dd0768   104857600   xfs    BAD      OPERATIVE   PUBLISHED   RELEASED   433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   6ada4e6d-861a-4454-8c86-71d24be4f992
$ drives
Id                                     Size        Type   SN                   Health   Usage      Status   Node
433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   105906176   HDD    LOOPBACK3898502982   BAD      RELEASED   ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dc5e6b7-3f38-4515-b420-5d003d614e34   104857600   HDD    LOOPBACK2176849526   GOOD     IN_USE     ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dcbd70a-02d6-4f25-96cc-1c4f66932972   104857600   HDD    LOOPBACK4005496940   GOOD     IN_USE     ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
```
- Mark _drive_ as ready for replacement
```
$ kubectl annotate drive 433ea9b0-d10b-42a0-ab6a-9d3862e2bd93 replacement=ready
drive.baremetal-csi.dellemc.com/433ea9b0-d10b-42a0-ab6a-9d3862e2bd93 annotated
$ drives
Id                                     Size        Type   SN                   Health   Usage      Status   Node
433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   105906176   HDD    LOOPBACK3898502982   BAD      REMOVING   ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dc5e6b7-3f38-4515-b420-5d003d614e34   104857600   HDD    LOOPBACK2176849526   GOOD     IN_USE     ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dcbd70a-02d6-4f25-96cc-1c4f66932972   104857600   HDD    LOOPBACK4005496940   GOOD     IN_USE     ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
```
- Add new drive and verify that replacement completed successfully
```
# add new drive
$ cat loopback.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: default
  name: loopback-config
  labels:
    app: baremetal-csi-node
data:
  config.yaml: |-
    defaultDrivePerNodeCount: 4
    defaultDriveSize: 100Mi
    nodes:
    - nodeID: kind-worker 
      drives:
        - serialNumber: LOOPBACK3898502982 
          health: BAD

$ kubectl apply -f loopback.yaml 
configmap/loopback-config configured

$ drives
Id                                     Size        Type   SN                   Health   Usage     Status   Node
433ea9b0-d10b-42a0-ab6a-9d3862e2bd93   104857600   HDD    LOOPBACK3898502982   BAD      REMOVED   ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dc5e6b7-3f38-4515-b420-5d003d614e34   104857600   HDD    LOOPBACK2176849526   GOOD     IN_USE    ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
6dcbd70a-02d6-4f25-96cc-1c4f66932972   104857600   HDD    LOOPBACK4005496940   GOOD     IN_USE    ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992
f8d6df57-2ede-4dfb-aa9b-6eb8db9e3929   104857600   HDD    LOOPBACK332146017    GOOD     IN_USE    ONLINE   6ada4e6d-861a-4454-8c86-71d24be4f992

$ volumes
Id                                         Size        Type   Health   OpStatus    CSIStatus   Usage    Location                               Node
pvc-3d5078a7-e362-4de0-872c-274d6d75c0e3   104857600   xfs    GOOD     OPERATIVE   CREATED     IN_USE   6dc5e6b7-3f38-4515-b420-5d003d614e34   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-48a07884-b021-4b0e-a270-b714c30f9a03   104857600   xfs    GOOD     OPERATIVE   CREATED     IN_USE   6dcbd70a-02d6-4f25-96cc-1c4f66932972   6ada4e6d-861a-4454-8c86-71d24be4f992
pvc-b324e193-8a48-47b7-ae05-b4aca3639599   104857600   xfs    GOOD     OPERATIVE   CREATED     IN_USE   f8d6df57-2ede-4dfb-aa9b-6eb8db9e3929   6ada4e6d-861a-4454-8c86-71d24be4f992

$ kubectl get pvc
NAME         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
data-web-0   Bound    pvc-3d5078a7-e362-4de0-872c-274d6d75c0e3   100Mi      RWO            baremetal-csi-sc   10m
logs-web-0   Bound    pvc-b324e193-8a48-47b7-ae05-b4aca3639599   100Mi      RWO            baremetal-csi-sc   2m54s
www-web-0    Bound    pvc-48a07884-b021-4b0e-a270-b714c30f9a03   100Mi      RWO            baremetal-csi-sc   10m

$ kubectl get pods
NAME                                        READY   STATUS    RESTARTS   AGE
web-0                                       1/1     Running   0          42s
```

## PR checklist
- [x] Add link to the issue
- [ ] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved
- [ ] PR validation is passed
- [ ] Custom CI is passed

## Testing
_In progress_

### CODE REVIEW

File: pkg/node/volumemgr.go [hunk 1]
```diff
@@ -222,6 +222,21 @@ func (m *VolumeManager) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+			// we need to update annotation on related drive CRD
+			// todo can we do polling instead?
+			ll.Infof("Volume %s is removed. Updating related", volume.Name)
+			drive := m.crHelper.GetDriveCRByUUID(volume.Spec.Location)
+			annotationKey := fmt.Sprintf("volume/%s", volume.Name)
+			// init map if empty
+			if drive.Annotations == nil {
+				drive.Annotations = make(map[string]string)
+			}
+			drive.Annotations[annotationKey] = apiV1.Removed
```
Comment (AnastasiaShemyakinskaya) [inline pkg/node/volumemgr.go:234]: Why do we need to annotate drive CR?

Comment (mishoyama) [reply pkg/node/volumemgr.go:234]: To trigger reconcile method for this drive instance.

File: pkg/crcontrollers/drive/drivecontroller.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import (
+	"context"
+	"time"
+
+	apiV1 "github.com/dell/csi-baremetal/api/v1"
+	"github.com/dell/csi-baremetal/api/v1/drivecrd"
+	"github.com/dell/csi-baremetal/pkg/base/k8s"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
```
[the remark this answers is not in the document]
Comment (mishoyama) [reply pkg/crcontrollers/drive/drivecontroller.go:8]: Fixed.

**Review by AnastasiaShemyakinskaya — APPROVED** [review]

**Review by ykulazhenkov — APPROVED** [review]

File: pkg/crcontrollers/drive/drivecontroller.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		if volume == nil || volume.Spec.CSIStatus == apiV1.Removed {
+			// TODO need to call drive manager to start LED locate
+			drive.Spec.Usage = apiV1.DriveUsageRemoved
+			break
+		}
+		isChanged = false
+	}
+
+	// update drive CR if needed
+	if isChanged {
```
Comment (painhardcore) [inline pkg/crcontrollers/drive/drivecontroller.go:131]: Why do choose to toggle isChanged as true by the start?
Using break and ` isChanged = false` in every "case" is confusing. Why don't we mark ` isChanged = true` when it's actually changed

**Review by painhardcore — CHANGES_REQUESTED** [review]

File: pkg/crcontrollers/drive/drivecontroller.go (hunk 3 above)
Comment (mishoyama) [reply pkg/crcontrollers/drive/drivecontroller.go:131]: Ok, will refactor.

**Review by painhardcore — APPROVED** [review]

**Pull request merged**